### PR TITLE
Support distinct error type in balo and jvm codegen

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
$subject.

Support exporting distinct error type information in balo format and add preliminary runtime typechecking support to distinct error types.

Fixes #<Issue Number>

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BPackage.java [hunk 1]
```diff
@@ -35,18 +35,20 @@
...
         this.name = name;
         this.version = version;
+        hashCode = Objects.hash(org, name, version);
     }
 
     public BPackage(String org, String name) {
         this.org = org;
         this.name = name;
         this.version = "";
+        hashCode = Objects.hash(org, name);
```
Comment (hasithaa) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BPackage.java:51]: With Type ID, this should contain a version.

Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BPackage.java:51]: I didn't quite get you, for this particular constructor version is a constant right?

Comment (hasithaa) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BPackage.java:51]: One assumption in typeID design is, modules always have a version. Maybe we should remove this constructor.

Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BPackage.java:51]: Ack. Will look into removing this in a subsequent PR.

**Pull request merged**